src/parsers/simple-fasta-parser.c

```diff
@@ -0,0 +1,129 @@
+    sight[0] = 1;
+    int j = 1;
+    for(int k=FPA; k<LPA; k++) {
+        if(bigAlphabet[k]) {
+            sight[j] = bigAlphabet[k];
+            bigAlphabet[k] = j++;
+        }
+    }
+    return sight;
+}
+
+
+char *read_fasta_head(char **strptr) {
+    char *string = *strptr;
+    // We assume first character is '>', so skip it
+    string++;
+    // trim leading spaces
+    while (isspace(*string)) {
+        string++;
+    }
+    // Our header starts from this point
+    char* header_start = string;
+    // Move forward until a newline is encountered
+    int i = 0;
+    while (string[i] != '\n') {
+        i++;
+    }
+
+    // Go back until first non-space character and thus trim trailing spaces and the \n
+    int j = i;
+    while (isspace(header_start[j])) header_start[j--] = '\0';
+    // Move the pointer of the string to start of header plus count of all characters in header plus 1
+    *strptr = header_start+i+1;
+    // Now the start of the string points to the first sequence character, and we can return the FASTA header
+    return header_start;
+}
+
+void update_fasta_by_sequence(char **strptr, struct Fasta *f) {
+    // Space for all printable ascii chars, where each entry corresponds to how many times we've seen a character
+    int *bigAlphabet = calloc(LPA, sizeof(int));
+    // alphabetSize is for example 5 for ACGT (+1 for sentinel)
+    int alphabetSize = 1;
+
+    char *string = *strptr;
+    int i = 0;
+    int shift = 0;
+    while (1) {
+        // Increase shift for each \n we encounter
+        while (string[i+shift] == '\n' || string[i+shift] == '\r') {
+            shift++;
+        }
+        // If we encounter a \0 or >, then we are done
+        if (string[i+shift] == '\0' || string[i+shift] == '>') {
+            break;
+        }
+        // Look ahead and store c
+        char c = string[i+shift];
+        string[i] = c; // Apply the shift
+        if(!(bigAlphabet[c]++)) alphabetSize++; // If first time seen symbol, add to size
+        i++;
+    }
+    // Terminate with \0 and move the pointer to the first instance of >
+    string[i] = 0;
+    *strptr += i+shift;
+    int *sightings = alloc_sightings(bigAlphabet, alphabetSize);
+
+    char * debugger = malloc((i+1)*sizeof *debugger);
+    debugger[i+1] = '\0';
+
+    for(int l=0; l<i+1; l++) {
+        debugger[l] = string[l];
+        string[l] = (char) bigAlphabet[string[l]];
+    }
+
+    f->alphabet.size = alphabetSize;
+    f->alphabet.symbols = bigAlphabet;
+    f->alphabet.sightings = sightings;
+    f->fasta_sequence = string;
+    f->fasta_len = i+1;
+    f->fasta_sequence_debugger = debugger;
+}
+
+
+struct FastaContainer *parse_fasta(char *fasta_str) {
+    struct Fasta **fastas = malloc(magic_number*sizeof (**fastas));
+    int i = 0;
+    while (fasta_str[0] != '\0') {
+        char *header = read_fasta_head(&fasta_str);
+        struct Fasta *f = malloc(sizeof(*f));
+        f->fasta_head = header;
```

Comment:
Here, you allocate memory and put it in `f->fastsa_head`. Then I would expect that memory to be freed in `free_fastas`, but there you only free the record, `f` (when you do `free(fastas[i])`).

Memory management is *hard*, and it is a good idea to be structured about it. I would recommend that you have a `free_x()` for each structure you allocate, so that is responsible for freeing whatever something of type `x` is.

Here you are working with two structures

```c
struct Alphabet {
     int size;
     int *symbols;
     int *sightings;
 };

 struct Fasta {
     char *fasta_head;
     char *fasta_sequence;
     char *fasta_sequence_debugger;
     int fasta_len;
     struct Alphabet alphabet;
 };
```

I would write functions for freeing both:

```c
void free_alphabet(struct Alphabet *a)
{
  free(a->symbols);
  free(a->sightings);
  free(a);
}

void free_fasta(struct Fasta *f)
{
  free(f->fasta_head);
  free(f->fasta_sequence);
  free(f->fasta_sequence_debugger);
  free_alphabet(f->alphabet); // WARNING (see below)
}
```

Since you have chosen to embed the alphabet in the `Fasta` structure this doesn't work, and that might be the reason you don't do it. If you do want to embed the structure you can change the rule and say that the `free_x()` functions do not `free()` the structure itself, just what it is holding. Just come up with something consistent.

The reason that it is a good idea to have such functions is that you only need one place to change the memory management when you add things to a structure.

You can take this further and make it polymorphic. You can also easily add reference counting if you want. There are a ton of ways that you can manage memory, but the key point is that you have to find a way to do it consistently, or you are guaranteed to leak.